bson/bsoncodec/empty_interface_codec.go

```diff
@@ -57,13 +57,20 @@ func (eic EmptyInterfaceCodec) EncodeValue(ec EncodeContext, vw bsonrw.ValueWrit
 
 func (eic EmptyInterfaceCodec) getEmptyInterfaceDecodeType(dc DecodeContext, valueType bsontype.Type) (reflect.Type, error) {
 	isDocument := valueType == bsontype.Type(0) || valueType == bsontype.EmbeddedDocument
-	if isDocument && dc.Ancestor != nil {
+	if isDocument && dc.HasValidAncestor() {
```

Comment:
Instead of having a separate method for this check, could we just do:
```
if isDocument {
    if dc.DocumentDecodeType != nil {
        // If the bsontype is an embedded document and the DocumentDecodeType is set on the DecodeContext, then just
	// return that type.  This gives users an "escape hatch" to avoid the 1-1 embedded document to primitive.D
	// decoding map, but this solution is AON.
        return *dc.DocumentDecodeType, nil
    }
    if dc.Ancestor != nil {
        // In the absence of a provided DocumentDecodeType, using ancestor information rather than looking up the type map entry forces consistent decoding.
	// If we're decoding into a bson.D, subdocuments should also be decoded as bson.D, even if a type map entry
	// has been registered.
        return dc.Ancestor, nil
    }
}
```